File: Data Structures & Algorithms/linked-list-cycle-detection/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    bool hasCycle(ListNode* head) {
        // key = address
        // value = "index" of key
        unordered_map<ListNode*, int> visited;
        ListNode* curr = head;
        int index = -1; // used to track index

        while (curr != nullptr) {
            // If this node hasn't been visited, mark where it was seen
            if (visited.contains(curr)) {
                return true;
            }
            else {
                ++index;
                visited[curr] = index;
            }

            curr = curr->next;
        }

        return false;
    }
};

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-1.py
import heapq

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:    
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        min_heap = []
        unique_id = 0

        for head in lists:
            if head is not None:
                heapq.heappush(min_heap, (head.val, unique_id, head))
                unique_id += 1

        dummy = ListNode(0)
        curr = dummy

        while min_heap:
            value, _, smallest = heapq.heappop(min_heap)

            curr.next = smallest
            curr = curr.next

            if smallest.next is not None:
                heapq.heappush(
                    min_heap,
                    (smallest.next.val, unique_id, smallest.next)
                )
                unique_id += 1

        return dummy.next



        

